segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkManager.java

```diff
@@ -0,0 +1,1308 @@
+     * to leverage multi-part uploads in object stores that support it (e.g., AWS S3, Dell EMC ECS) as they are typically
+     * only metadata operations, reducing the overall cost of the merging them together. HDFS also supports merges,
+     * whereas NFS has no concept of merging natively.
+     *
+     * As chunks become larger, append writes (read source completely and append-i.e., write- it back at the end of target)
+     * become inefficient. Consequently, a native option for merging is desirable. We use such native merge capability
+     * when available, and if not available, then we use appends.
+     * </li>
+     * <li>
+     * Ideally we want the defrag to be run in the background periodically and not on the write/concat path.
+     * We can then fine tune that background task to run optimally with low overhead.
+     * We might be able to give more knobs to tune its parameters (Eg. threshold on number of chunks).
+     * </li>
+     * <li>
+     * <li>
+     * Defrag operation will respect max rolling size and will not create chunks greater than that size.
+     * </li>
+     * </ul>
+     *
+     * <div>
+     * What controls whether we invoke concat or simulate through appends?
+     * There are a few different capabilities that ChunkStorage needs to provide.
+     * <ul>
+     * <li>Does ChunkStorage support appending to existing chunks? For vanilla S3 compatible this would return false.
+     * This is indicated by supportsAppend.</li>
+     * <li>Does ChunkStorage support for concatenating chunks ? This is indicated by supportsConcat.
+     * If this is true then concat operation will be invoked otherwise chunks will be appended.</li>
+     * <li>There are some obvious constraints - For ChunkStorage support any concat functionality it must support either
+     * append or concat.</li>
+     * <li>Also when ChunkStorage supports both concat and append, ChunkManager will invoke appropriate method
+     * depending on size of target and source chunks. (Eg. ECS)</li>
+     * </ul>
+     * </div>
+     * <li>
+     * What controls defrag?
+     * There are two additional parameters that control when concat
+     * <li>minSizeLimitForConcat: Size of chunk in bytes above which it is no longer considered a small object.
+     * For small source objects, append is used instead of using concat. (For really small txn it is rather efficient to use append than MPU).</li>
+     * <li>maxSizeLimitForConcat: Size of chunk in bytes above which it is no longer considered for concat. (Eg S3 might have max limit on chunk size).</li>
+     * In short there is a size beyond which using append is not advisable. Conversely there is a size below which concat is not efficient.(minSizeLimitForConcat )
+     * Then there is limit which concating does not make sense maxSizeLimitForConcat
+     * </li>
+     * <li>
+     * What is the defrag algorithm
+     * <pre>
+     * While(segment.hasConcatableChunks()){
```

Comment:
If I'm not mistaken, you are executing this logic as part of `concat()` and, in addition to that, you are executing `collectGarbage()` as part of `write()`. These are the two most performance critical operations, as they will dictate the write throughput and the transaction throughput to Tier 2, respectively. Please, consider if these 2 tasks should be executed on a regular basis by an independent service instead of adding this burden to the regular write activity of Tier 2.

Reply:
There is a planned `StorageOptimiser` component that is planned but not part of this set of PRs. https://github.com/pravega/pravega/wiki/PDP-34:-Simplified-Tier-2#components
https://github.com/pravega/pravega/issues/4903